File: core/src/main/java/com/oneandone/typedrest/ConsumerEndpointImpl.java
package com.oneandone.typedrest;

import java.io.*;
import java.net.*;
import lombok.Getter;
import org.apache.http.client.fluent.*;
import org.apache.http.entity.*;

/**
 * REST endpoint that represents an RPC-like action which takes
 * <code>TEntity</code> as input.
 *
 * @param <TEntity> The type of entity the endpoint takes as input.
 */
public class ConsumerEndpointImpl<TEntity>
        extends AbstractTriggerEndpoint
        implements ConsumerEndpoint<TEntity> {

    @Getter
    private final Class<TEntity> entityType;

    /**
     * Creates a new action endpoint with a relative URI.
     *
     * @param parent The parent endpoint containing this one.
     * @param relativeUri The URI of this endpoint relative to the
     * <code>parent</code>'s.
     * @param entityType The type of entity the endpoint takes as input.
     */
    public ConsumerEndpointImpl(Endpoint parent, URI relativeUri, Class<TEntity> entityType) {
        super(parent, relativeUri);
        this.entityType = entityType;
    }

    /**
     * Creates a new action endpoint with a relative URI.
     *
     * @param parent The parent endpoint containing this one.
     * @param relativeUri The URI of this endpoint relative to the
     * <code>parent</code>'s. Prefix <code>./</code> to append a trailing slash
     * to the parent URI if missing.
     * @param entityType The type of entity the endpoint takes as input.
     */
    public ConsumerEndpointImpl(Endpoint parent, String relativeUri, Class<TEntity> entityType) {
        super(parent, relativeUri);
        this.entityType = entityType;
    }

    @Override
    public void trigger(TEntity entity)
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException {
        if (entity == null) {
            throw new IllegalArgumentException("entity must not be null.");
        }

        String jsonSend = serializer.writeValueAsString(entity);
        executeAndHandle(Request.Put(uri).bodyString(jsonSend, ContentType.APPLICATION_JSON));
    }
}

File: typedrest/src/main/kotlin/net/typedrest/endpoints/rpc/ConsumerEndpointImpl.kt
package net.typedrest.endpoints.rpc

import net.typedrest.endpoints.*
import net.typedrest.http.uri
import okhttp3.Request
import java.net.URI

/**
 * RPC endpoint that takes [TEntity] as input when invoked.
 *
 * @param referrer The endpoint used to navigate to this one.
 * @param relativeUri The URI of this endpoint relative to the [referrer]'s. Add a `./` prefix here to imply a trailing slash on referrer's URI.
 * @param entityType The type of entity the endpoint takes as input.
 * @param TEntity The type of entity the endpoint takes as input.
 */
open class ConsumerEndpointImpl<TEntity>(
    referrer: Endpoint,
    relativeUri: URI,
    private val entityType: Class<TEntity>
) : AbstractRpcEndpoint(referrer, relativeUri), ConsumerEndpoint<TEntity> {
    /**
     * Creates a new consumer endpoint.
     *
     * @param referrer The endpoint used to navigate to this one.
     * @param relativeUri The URI of this endpoint relative to the [referrer]'s. Add a `./` prefix here to imply a trailing slash on referrer's URI.
     * @param entityType The type of entity the endpoint takes as input.
     */
    constructor(referrer: Endpoint, relativeUri: String, entityType: Class<TEntity>) :
        this(referrer, URI(relativeUri), entityType)

    override fun invoke(entity: TEntity) =
        execute(Request.Builder().post(serialize(entity, entityType)).uri(uri).build()).close()
}
